src/spi/bus.rs

```diff
@@ -0,0 +1,369 @@
+    pub fn listen_rx_wm(&mut self) {
+        self.spi.ie.write(|w| w.rxwm().set_bit())
+    }
+
+    /// Stops listening for transmit watermark interrupt event
+    pub fn unlisten_tx_wm(&mut self) {
+        self.spi.ie.write(|w| w.txwm().clear_bit())
+    }
+
+    /// Stops listening for receive watermark interrupt event
+    pub fn unlisten_rx_wm(&mut self) {
+        self.spi.ie.write(|w| w.rxwm().clear_bit())
+    }
+
+    /// Set AUTO CS mode to per-word operation, unless CSMODE is set to OFF
+    pub fn cs_mode_word(&mut self) {
+        if !self.spi.csmode.read().mode().is_off() {
+            self.spi.csmode.write(|w| w.mode().auto());
+        }
+    }
+
+    /// Set HOLD CS mode to per-frame operation, unless CSMODE is set to OFF
+    pub fn cs_mode_frame(&mut self) {
+        if !self.spi.csmode.read().mode().is_off() {
+            self.spi.csmode.write(|w| w.mode().hold());
+        }
+    }
+
+    /// Finishes transfer by deasserting CS (only for hardware-controlled CS)
+    pub fn end_transfer(&mut self) {
+        self.cs_mode_word()
+    }
+
+    /// Releases the SPI peripheral and associated pins
+    pub fn free(self) -> (SPI, PINS) {
+        (self.spi, self.pins)
+    }
+
+    /// Configure the SPI bus for given mode, frequency and CS index
+    pub(crate) fn configure(
+        &mut self,
+        mode: Mode,
+        clock_divisor: u32,
+        cs_mode: e310x::qspi0::csmode::MODE_A,
+        cs_index: Option<u32>,
+    ) where
+        PINS: Pins<SPI>,
+    {
+        assert!(clock_divisor <= 0xfff);
+        self.spi
+            .sckdiv
+            .write(|w| unsafe { w.div().bits(clock_divisor as u16) });
+
+        if let Some(index) = cs_index {
+            self.spi.csid.write(|w| unsafe { w.bits(index) });
+        }
+        self.spi.csmode.write(|w| w.mode().variant(cs_mode));
+
+        // Set CS pin polarity to high
+        self.spi.csdef.reset();
+
+        // Set SPI mode
+        let phase = mode.phase == Phase::CaptureOnSecondTransition;
+        let polarity = mode.polarity == Polarity::IdleHigh;
+        self.spi
+            .sckmode
+            .write(|w| w.pha().bit(phase).pol().bit(polarity));
+
+        self.spi.fmt.write(|w| unsafe {
+            w.proto().single();
+            w.endian().big(); // Transmit most-significant bit (MSB) first
+            w.dir().rx();
+            w.len().bits(8)
+        });
+
+        // Set watermark levels
+        self.spi.txmark.write(|w| unsafe { w.txmark().bits(1) });
+        self.spi.rxmark.write(|w| unsafe { w.rxmark().bits(0) });
+
+        // set delays to their reset values
+        self.spi.delay0.write(|w| unsafe {
+            w.cssck().bits(1); // no delay on CS assert
+            w.sckcs().bits(1) // 1 SCK cycle delay on CS deassert
+        });
+        self.spi.delay1.write(|w| unsafe {
+            w.intercs().bits(1); // min 1 SCK cycle between CS re-assets
+            w.interxfr().bits(0) // no intra-frame delay
+        });
```

Comment:
Probably we need to set all-0 values, otherwise SPI works in a weird way. You know, it worked before because of an absent reset value in PAC.

Reply:
I think for shared mode we might need the proper reset value and for exclusive it doesn't matter as the HOLD/CS is done manually anyhow. I'll have to check this closer to be sure. I think the devices might get into "both registering" if the delay is missing.

Reply:
IIRC, even in exclusive mode this 1-1-1-0 setting will de-assert CS between bytes and this is something we definitely do not want.

Reply:
seems to be that way